include/nbl/builtin/hlsl/subgroup/arithmetic_portability_impl.hlsl

```diff
@@ -0,0 +1,490 @@
+		}
+		Barrier();
+		MemoryBarrierShared();
+		
+		// Stone-Kogge adder
+		// (devsh): it seems that lanes below <HalfSubgroupSize/step are doing useless work,
+		// but they're SIMD and adding an `if`/conditional execution is more expensive
+	#if 1 // Use shuffling by default (either native or portability implementation)
+		uint toAdd = ShuffleUp<T, ScratchAccessor>(value, 1u); // all invocations must execute the shuffle, even if we don't apply the op() to all of them
+		if(offsetsAndMasks.subgroupInvocation >= 1u) {
+			// the first invocation (index 0) in the subgroup doesn't have anything in its left
+			value = op(value, toAdd);
+		}
+	#else
+		value = op(value, scratch.main.get(offsetsAndMasks.scanStoreOffset-1u));
+	#endif
+		[[unroll]] // REVIEW: I copied this from the example implementation on the github issue but if I'm not mistaken unroll doesn't work since halfSubgroupSize is not constexpr 
+		for (uint step=2u; step <= offsetsAndMasks.halfSubgroupSize; step <<= 1u)
+		{
+		#if 1
+			// there is no scratch and padding entries in this case so we have to guard the shuffles to not go out of bounds
+			toAdd = ShuffleUp<T, ScratchAccessor>(value, step);
+			if(offsetsAndMasks.subgroupInvocation >= step) {
+				value = op(value, toAdd);
+			}
+		#else
+			Barrier();
+			MemoryBarrierShared();
+			scratch.main.set(offsetsAndMasks.scanStoreOffset, value);
+			Barrier();
+			MemoryBarrierShared();
+			value = op(value, scratch.main.get(offsetsAndMasks.scanStoreOffset - step));
+			Barrier();
+			MemoryBarrierShared();
+			scratch.main.set(offsetsAndMasks.scanStoreOffset, value);
+			Barrier();
+			MemoryBarrierShared();
+			
+			// REVIEW: The Stone-Kogge adder is done at this point.
+			// The GLSL implementation however has a final operation that uses the lastLoadOffset
+			// but it actually messes the results, not sure what the point was.
+			//value = op(value, scratch.main.get(offsetsAndMasks.lastLoadOffset));
+			//Barrier();
+			//MemoryBarrierShared();
+		#endif
+		}
+		return value;
+    }
+// protected:
+	ScratchAccessor scratch;
+	ScratchOffsetsAndMasks offsetsAndMasks;
+};
+
+template<typename T, class Binop, class ScratchAccessor, bool initializeScratch>
+struct exclusive_scan
+{
+    static exclusive_scan<T, Binop, ScratchAccessor, initializeScratch> create()
+    {
+        exclusive_scan<T, Binop, ScratchAccessor, initializeScratch> retval;
+        retval.impl = inclusive_scan<T, Binop, ScratchAccessor, initializeScratch>::create();
+        return retval;
+    }
+
+    T operator()(T value)
+    {
+		value = impl(value);
+
+		// store value to smem so we can shuffle it
+	#if 1
+		Binop op;
+		uint left = ShuffleUp<T, ScratchAccessor>(value, 1);
+		value = impl.offsetsAndMasks.subgroupInvocation >= 1 ? left : op.identity(); // the first invocation doesn't have anything in its left so we set to the binop's identity value for exlusive scan
```

Comment:
you can actually shuffle before or after inclusive scanning, benchmark what's faster

Reply:
verdict?

Reply:
ok according to Discord priv convo from today, this is no longer relevant after redesign.